core/src/main/java/com/datastax/oss/driver/api/core/session/Session.java

```diff
@@ -54,101 +60,76 @@
   CqlIdentifier getKeyspace();
 
   /**
-   * Executes a request, and blocks until the result is available.
+   * Executes an arbitrary request.
    *
-   * @return a synchronous result, that provides immediate access to the data as soon as the method
-   *     returns.
+   * @param resultType the type of the result, which determines the internal request processor
+   *     (built-in or custom) that will be used to handle the request.
+   * @see Session
    */
-  <SyncResultT, AsyncResultT> SyncResultT execute(Request<SyncResultT, AsyncResultT> request);
+  <RequestT extends Request, ResultT> ResultT execute(
+      RequestT request, GenericType<ResultT> resultType);
 
   /**
-   * Executes a request, returning as soon as it has been scheduled, but generally before the result
-   * is available.
-   *
-   * @return an asynchronous result, that represents the future completion of the request. The
-   *     client either wait, or schedule a callback to be executed on completion (this is
-   *     implementation-specific).
-   */
-  <SyncResultT, AsyncResultT> AsyncResultT executeAsync(Request<SyncResultT, AsyncResultT> request);
-
-  /**
-   * Executes a CQL statement synchronously.
-   *
-   * <p>This is a convenience method that does the exact same thing as {@link #execute(Request)},
-   * but exposes a more user-friendly signature reminiscent of the 3.x API.
+   * Executes a CQL statement synchronously (the calling thread blocks until the result becomes
+   * available).
    */
   default ResultSet execute(Statement<?> statement) {
```

Comment:
This is not completely related to the PR, but I had a thought about these `default` methods.  If I'm implementing my own `Session` implementation that simply wraps `Session` and provides some extra sugar methods wouldn't they clash with these?

Say for example I wanted to make a `GuavaSession` which provides async methods for returning `ListenableFuture`, the `executeAsync` methods from this interface come along with it, so I try to define `ListenableFuture<ResultSet> executeAsync(Statement<?> statement)`.  This would not work because the signature would clash with `CompletionStage<ResultSet> executeAsync(Statement<?> statement)`.

Do you see this as being a problem?  One possible solution would be to break Session into two interfaces.  A top level one (still named `Session`) that only has `getKeyspace` and the `execute` method that requires `Request` and then another one that extends `Session` and provides these specialized default methods (`DefaultSession` would extend this).

I mostly started thinking about this due to the general `execute` method taking a `Request`.  I figure library implementors will want to write customized sessions that provide sugar methods like these for the applicable types, but maybe the approach would be to make something that isn't a `Session`, but uses one and calls this method on the behalf of the user.

If the approach is to have implementors come up with their own `Session`.  It would be nice to generalize `Cluster` with its `connect` methods returning a type of `Session`.

I think I will just need to try implementing something to wrap my head around it.  I think this is the best approach to reviewing this (try implementing something that uses it).

Reply:
I was revisiting some of our previous discussion and I think our thought was not to make specific `Session` implementations as things can get out of hand quickly, so maybe the *make something that isn't a Session, but uses one and calls this method on the behalf of the user* is the best way to go.

Reply:
I can't think of a good solution to this in Java. I really think wrapping is not the solution, the main problem is that it doesn't compose.

What we would need is a way to dynamically mix in the different capabilities, returning some kind of union type:
```
session = cluster.connect() with GuavaCqlMethods with ReactiveCqlMethods
```
Obviously in Java this is science fiction.

All things considered, passing a type token to every `Session.execute` call is not that bad once you got used to it.